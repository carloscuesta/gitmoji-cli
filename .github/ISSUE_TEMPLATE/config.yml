blank_issues_enabled: false
contact_links:
  - name: 😍 Contribution Guide
    url: https://github.com/carloscuesta/gitmoji-cli/blob/master/.github/CONTRIBUTING.md
    about: Please read through before making any contribution.