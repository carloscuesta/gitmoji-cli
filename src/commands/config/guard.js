import chalk from 'chalk'
import isURL from 'validator/lib/isURL.js'

const errors = {
  url: chalk.red('Enter a valid API URL')
}

const url = (url: string) => (isURL(url) ? true : errors.url)

export default {
  url
}
