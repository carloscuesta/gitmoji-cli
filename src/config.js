const Conf = require('conf')

const constants = require('./constants')
const config = new Conf()

const getAutoAdd = () => config.get(constants.AUTO_ADD)
const getEmojiFormat = () => config.get(constants.EMOJI_FORMAT)
const getIssueFormat = () => config.get(constants.ISSUE_FORMAT)
const getSignedCommit = () => config.get(constants.SIGNED_COMMIT)
const setAutoAdd = (autoAdd) => config.set(constants.AUTO_ADD, autoAdd)
const setEmojiFormat = (emojiFormat) => {
  config.set(constants.EMOJI_FORMAT, emojiFormat)
}
const setIssueFormat = (issueFormat) => {
  config.set(constants.ISSUE_FORMAT, issueFormat)
}
const setSignedCommit = (signedCommit) => {
  config.set(constants.SIGNED_COMMIT, signedCommit)
}

module.exports = {
  getAutoAdd,
  getEmojiFormat,
  getIssueFormat,
  getSignedCommit,
  setAutoAdd,
  setEmojiFormat,
  setIssueFormat,
  setSignedCommit
}
