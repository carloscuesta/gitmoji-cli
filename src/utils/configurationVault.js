// @flow
import Conf from 'conf'

import {
  CONFIGURATION_PROMPT_NAMES,
  EMOJI_COMMIT_FORMATS
} from '../commands/config/prompts'

export const config = new Conf({ projectName: 'gitmoji' })

const setAutoAdd = (autoAdd: boolean) => {
  config.set(CONFIGURATION_PROMPT_NAMES.AUTO_ADD, autoAdd)
}

const setEmojiFormat = (emojiFormat: string) => {
  config.set(CONFIGURATION_PROMPT_NAMES.EMOJI_FORMAT, emojiFormat)
}

const setSignedCommit = (signedCommit: boolean) => {
  config.set(CONFIGURATION_PROMPT_NAMES.SIGNED_COMMIT, signedCommit)
}

const setScopePrompt = (scopePrompt: boolean) => {
  config.set(CONFIGURATION_PROMPT_NAMES.SCOPE_PROMPT, scopePrompt)
}

const getAutoAdd = (): boolean => {
  return config.get(CONFIGURATION_PROMPT_NAMES.AUTO_ADD) || false
}

const getEmojiFormat = (): string => {
  return (
    config.get(CONFIGURATION_PROMPT_NAMES.EMOJI_FORMAT) ||
    EMOJI_COMMIT_FORMATS.CODE
  )
}

const getSignedCommit = (): boolean => {
  return config.get(CONFIGURATION_PROMPT_NAMES.SIGNED_COMMIT) || false
}

const getScopePrompt = (): boolean => {
  return config.get(CONFIGURATION_PROMPT_NAMES.SCOPE_PROMPT) || false
}

export default {
  getAutoAdd,
  getEmojiFormat,
  getScopePrompt,
  getSignedCommit,
  setAutoAdd,
  setEmojiFormat,
  setScopePrompt,
  setSignedCommit
}
