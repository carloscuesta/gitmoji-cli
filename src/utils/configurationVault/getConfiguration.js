import Conf from 'conf'
import { cwd } from 'process'
import { readFileSync } from 'fs'
import { pathExistsSync } from 'path-exists'

import { CONFIG, EMOJI_COMMIT_FORMATS } from '@constants/configuration'

const DEFAULT_CONFIGURATION = {
  [CONFIG.AUTO_ADD]: false,
  [CONFIG.EMOJI_FORMAT]: EMOJI_COMMIT_FORMATS.CODE,
  [CONFIG.SCOPE_PROMPT]: false,
  [CONFIG.MESSAGE_PROMPT]: true,
  [CONFIG.CAPITALIZE_TITLE]: true,
  [CONFIG.GITMOJIS_URL]: 'https://gitmoji.dev/api/gitmojis'
}

const LOCAL_CONFIGURATION: typeof Conf = new Conf({
  projectName: 'gitmoji',
  schema: {
    [CONFIG.AUTO_ADD]: {
      type: 'boolean',
      default: DEFAULT_CONFIGURATION[CONFIG.AUTO_ADD]
    },
    [CONFIG.EMOJI_FORMAT]: {
      enum: Object.values(EMOJI_COMMIT_FORMATS),
      default: DEFAULT_CONFIGURATION[CONFIG.EMOJI_FORMAT]
    },
    [CONFIG.SCOPE_PROMPT]: {
      type: 'boolean',
      default: DEFAULT_CONFIGURATION[CONFIG.SCOPE_PROMPT]
    },
    [CONFIG.MESSAGE_PROMPT]: {
      type: 'boolean',
      default: DEFAULT_CONFIGURATION[CONFIG.MESSAGE_PROMPT]
    },
    [CONFIG.CAPITALIZE_TITLE]: {
      type: 'boolean',
      default: DEFAULT_CONFIGURATION[CONFIG.CAPITALIZE_TITLE]
    },
    [CONFIG.GITMOJIS_URL]: {
      type: 'string',
      format: 'url',
      default: DEFAULT_CONFIGURATION[CONFIG.GITMOJIS_URL]
    }
  }
})

const getFile = (path: string): Buffer | void => {
  try {
    return JSON.parse(readFileSync(path))
  } catch (error) {
    return
  }
}

const getConfiguration = (): { get: Function, set: Function } => {
  const loadConfig = (): {
    [$Values<typeof CONFIG>]: string | boolean
  } => {
    const packageJson = `${cwd()}/package.json`
    const configurationFile = `${cwd()}/.gitmojirc.json`

    if (pathExistsSync(packageJson) && getFile(packageJson)?.gitmoji) {
      return getFile(packageJson)?.gitmoji
    }

    if (pathExistsSync(configurationFile) && getFile(configurationFile)) {
      return getFile(configurationFile)
    }

    return LOCAL_CONFIGURATION.store
  }

  return {
    get: (key: string): string | boolean => {
      const resolvedConfiguration = loadConfig()
      const configuration =
        typeof resolvedConfiguration === 'object' &&
        Object.keys(resolvedConfiguration).length
          ? resolvedConfiguration
          : DEFAULT_CONFIGURATION

      return configuration[key] ?? DEFAULT_CONFIGURATION[key]
    },
    set: (key: string, value: string | boolean): void =>
      LOCAL_CONFIGURATION.set(key, value)
  }
}

export default getConfiguration
