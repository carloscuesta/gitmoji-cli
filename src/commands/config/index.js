// @flow
import inquirer from 'inquirer'

import configurationPrompts, { CONFIGURATION_PROMPT_NAMES } from './prompts'
import configurationVault from '../../utils/configurationVault'

const config = () => {
  inquirer.prompt(configurationPrompts()).then((answers) => {
    configurationVault.setAutoAdd(answers[CONFIGURATION_PROMPT_NAMES.AUTO_ADD])
    configurationVault.setEmojiFormat(
      answers[CONFIGURATION_PROMPT_NAMES.EMOJI_FORMAT]
    )
    configurationVault.setSignedCommit(
      answers[CONFIGURATION_PROMPT_NAMES.SIGNED_COMMIT]
    )
    configurationVault.setScopePrompt(
      answers[CONFIGURATION_PROMPT_NAMES.SCOPE_PROMPT]
    )
  })
}

export default config
