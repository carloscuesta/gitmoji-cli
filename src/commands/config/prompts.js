// @flow
import configurationVault from '../../utils/configurationVault'

export const CONFIGURATION_PROMPT_NAMES = {
  AUTO_ADD: 'autoAdd',
  EMOJI_FORMAT: 'emojiFormat',
  SCOPE_PROMPT: 'scopePrompt',
  SIGNED_COMMIT: 'signedCommit'
}

export const EMOJI_COMMIT_FORMATS = {
  CODE: 'code',
  EMOJI: 'emoji'
}

export default () => [
  {
    name: CONFIGURATION_PROMPT_NAMES.AUTO_ADD,
    message: 'Enable automatic "git add ."',
    type: 'confirm',
    default: configurationVault.getAutoAdd()
  },
  {
    name: CONFIGURATION_PROMPT_NAMES.EMOJI_FORMAT,
    message: 'Select how emojis should be used in commits',
    type: 'list',
    choices: [
      { name: ':smile:', value: EMOJI_COMMIT_FORMATS.CODE },
      { name: 'ðŸ˜„', value: EMOJI_COMMIT_FORMATS.EMOJI }
    ],
    default: configurationVault.getEmojiFormat()
  },
  {
    name: CONFIGURATION_PROMPT_NAMES.SIGNED_COMMIT,
    message: 'Enable signed commits',
    type: 'confirm',
    default: configurationVault.getSignedCommit()
  },
  {
    name: CONFIGURATION_PROMPT_NAMES.SCOPE_PROMPT,
    message: 'Enable scope prompt',
    type: 'confirm',
    default: configurationVault.getScopePrompt()
  }
]
