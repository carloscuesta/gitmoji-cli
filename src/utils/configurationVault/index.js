// @flow
import { CONFIG, EMOJI_COMMIT_FORMATS } from '@constants/configuration'
import getConfiguration from './getConfiguration'

const configuration = getConfiguration()

const setAutoAdd = (autoAdd: boolean): void => {
  configuration.set(CONFIG.AUTO_ADD, autoAdd)
}

const setEmojiFormat = (emojiFormat: string): void => {
  configuration.set(CONFIG.EMOJI_FORMAT, emojiFormat)
}

const setScopePrompt = (scopePrompt: boolean): void => {
  configuration.set(CONFIG.SCOPE_PROMPT, scopePrompt)
}

const setMessagePrompt = (messagePrompt: boolean): void => {
  configuration.set(CONFIG.MESSAGE_PROMPT, messagePrompt)
}

const setGitmojisUrl = (gitmojisUrl: string): void => {
  configuration.set(CONFIG.GITMOJIS_URL, gitmojisUrl)
}

const getAutoAdd = (): boolean => {
  return configuration.get(CONFIG.AUTO_ADD)
}

const getEmojiFormat = (): string => {
  return configuration.get(CONFIG.EMOJI_FORMAT)
}

const getScopePrompt = (): boolean => {
  return configuration.get(CONFIG.SCOPE_PROMPT)
}

const getMessagePrompt = (): boolean => {
  return configuration.get(CONFIG.MESSAGE_PROMPT)
}

const getGitmojisUrl = (): string => {
  return configuration.get(CONFIG.GITMOJIS_URL)
}

export default {
  getAutoAdd,
  getEmojiFormat,
  getScopePrompt,
  getMessagePrompt,
  getGitmojisUrl,
  setAutoAdd,
  setEmojiFormat,
  setScopePrompt,
  setMessagePrompt,
  setGitmojisUrl
}
