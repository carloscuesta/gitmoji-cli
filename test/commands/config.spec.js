import inquirer from 'inquirer'

import config from '../../src/commands/config'
import configurationPrompts from '../../src/commands/config/prompts'
import configurationVault from '../../src/utils/configurationVault'
import * as stubs from './stubs'

jest.mock('../../src/utils/configurationVault')

describe('config command', () => {
  beforeAll(() => {
    inquirer.prompt.mockReturnValue(Promise.resolve(stubs.configAnswers))
    config()
  })

  it('should call inquirer.prompt with configuration prompts', () => {
    expect(inquirer.prompt).toHaveBeenCalledWith(configurationPrompts())
  })

  it('should save the answers into the configuration vault', () => {
    expect(configurationVault.setAutoAdd).toHaveBeenCalledWith(
      stubs.configAnswers.autoAdd
    )
    expect(configurationVault.setEmojiFormat).toHaveBeenCalledWith(
      stubs.configAnswers.emojiFormat
    )
    expect(configurationVault.setSignedCommit).toHaveBeenCalledWith(
      stubs.configAnswers.signedCommit
    )
    expect(configurationVault.setScopePrompt).toHaveBeenCalledWith(
      stubs.configAnswers.scopePrompt
    )
  })
})
