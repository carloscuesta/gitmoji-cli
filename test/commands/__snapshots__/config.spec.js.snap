// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config command guard should match guard 1`] = `
{
  "url": [Function],
}
`;
